### DISCUSSION TITLE
Using '.run' and 'await in' to observe AnyPublisher<Value, Never> does not work

### DISCUSSION THREAD

**Discussion by SebastianBoldt** [pointfreeco/swift-composable-architecture#1922] [Q&A]
Hey, 

I currently do have some Issues related to long living effects.

To give you some background:
I created a Dependency to observe a specific value from the UserDefaults.
It returns an AnyPublisher we can subscribe to, to get notified when the value changes.

```swift 

enum CardMode: String, Codable, CaseIterable {
    case expanded
    case collapsed
}

struct CardModeProvider {
    var cardMode: () -> CardMode
    var cardModePublisher: AnyPublisher<CardMode, Never>
    var storeCardMode: (CardMode) -> Void
}

extension CardModeProvider {
    static let live: CardModeProvider = {
        return Self(
            cardMode: {
                CardMode(rawValue: UserDefaults.sharedGroupDefaults.cardMode) ?? .expanded
            },
            cardModePublisher: UserDefaults.sharedGroupDefaults
                .publisher(for: \.cardMode)
                .map { CardMode(rawValue: $0) ?? .expanded }
                .eraseToAnyPublisher(),
            storeCardMode: {
                UserDefaults.sharedGroupDefaults.cardMode = $0.rawValue
            }
        )
    }()
}

private enum CardModeDependencyKey: DependencyKey {
    static let liveValue = CardModeProvider.live
}

extension DependencyValues {
  var cardModeProvider: CardModeProvider {
    get { self[CardModeDependencyKey.self] }
    set { self[CardModeDependencyKey.self] = newValue }
  }
}
```

Inside the reducer I observed the value changes using eraseToEffect:

```swift
func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
    switch action {
        case .onAppear:
            return cardModeProvider.cardModePublisher.eraseToEffect {
                .updateCardMode($0)
            }
    }
}
```

it works like a charm but after I took a look into the Docs I discovered that it is recommended to use `.run` and `for await value in sequence` for that. So I tried to do that.

```swift 
.run { send in
    for await value in cardModeProvider.cardModePublisher.values {
        await send(.updateCardMode(value))
    }
}
```
But unfortunately I doesn't work as expected.
It will get triggered once, but each time I change the value in the User-Defaults  the action will not get send.
I first assumed that my publisher was cancelled or completed somehow under the hood so I tried to investigate using handleEvents.

```swift 
static let live: CardModeProvider = {
        return Self(
            cardMode: {
                CardMode(rawValue: UserDefaults.sharedGroupDefaults.cardMode) ?? .expanded
            },
            cardModePublisher: UserDefaults.sharedGroupDefaults
                .publisher(for: \.cardMode)
                .handleEvents(receiveCompletion: {
                    // ...
                }, receiveCancel: {
                    // ...
                })
                .map { CardMode(rawValue: $0) ?? .expanded }
                .eraseToAnyPublisher(),
            storeCardMode: {
                UserDefaults.sharedGroupDefaults.cardMode = $0.rawValue
            }
        )
  }()
 ```
But it turned out that none of those publisher events occur. 
Does anyone maybe has an idea what I am doing wrong?

**Comment by tgrapperon** [comment]
Hey @SebastianBoldt! If `UserDefaults.sharedGroupDefaults` is a stable instance (a singleton for example), the async loop you presented should produce new values/actions if the publisher is sound.
BTW, you have the `.print()` method that works on `Publisher`, an that shows various informations about the subscription. You can maybe try to loop over `cardModeProvider.cardModePublisher.print().values` to see if it emits or not.
I would also write a test for the `cardModeProvider` publisher, to confirm that it works as expected in isolation. This publisher is KVO-based, but it should work for `UserDefaults` (that's what I'm using in `swift-dependencies-additions`. 
If nothing stands out, I would try to perform read/write on the `MainActor` for example, as this is something that could be different between isolated testing and TCA.

**Reply by SebastianBoldt** [reply]
Yes it is. 
```swift
extension UserDefaults {
    static let sharedGroupDefaults: UserDefaults = UserDefaults(suiteName: IdentifierConstants.AppGroups.defaultSharedGroup)!
}
```

Unfortunatly print does not print anything to the console if I attach it to cardModeProvider.cardModePublisher.

**Reply by tgrapperon** [reply]
There is some finesse required with the `KeyPath` extraction and observation. Have you annotated your value with `@objc`? Are you using the correct `set(:forKey:)` setters (and not `setValue(:forKey:)`)?

**Reply by SebastianBoldt** [reply]
This is what I did.
```swift 
import Foundation

extension UserDefaults {
    @objc var cardMode: String {
        get {
            return string(forKey: IdentifierConstants.UserDefaults.cardMode) ?? CardMode.expanded.rawValue
        }
        set {
            set(newValue, forKey: IdentifierConstants.UserDefaults.cardMode)
        }
    }
}
```

If something would be wrong with that, the approach that uses eraseToEffect ...
```swift 
func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
    switch action {
        case .onAppear:
            return cardModeProvider.cardModePublisher.eraseToEffect {
                .updateCardMode($0)
            }
    }
}
```
 ... should also not work, right?

**Reply by tgrapperon** [reply]
I'm not able to find the documentation for this method, so I can't really say what the default options are, but I would try to pass `userDefaults.publisher(for:\.cardMode, options: [.initial, .new])` if it seems to only emits on subscription (that would be `[.initial]` only, which would be rather odd as a default). Is it working when unit-testing?

**Reply by SebastianBoldt** [reply]
```swift
userDefaults.publisher(for:\.cardMode, options: [.initial, .new])
```
does not change anything. 

Unit-testing works like a charm. It is not the most beautiful test but it gets the job done.
```swift
    func testCardModeProvider() async throws {
        var newValue: CardMode?

        let cardModeProvider = CardModeProvider.live
        let publisher = cardModeProvider.cardModePublisher
        let oldCardModel = cardModeProvider.cardMode()

        publisher.sink(receiveValue: { newCardMode in
            newValue = newCardMode
        })
        .store(in: &cancellables)
        cardModeProvider.storeCardMode(.expanded)
        try await Task.sleep(for: .seconds(1))
        XCTAssertEqual(newValue, CardMode.expanded)

        cardModeProvider.storeCardMode(.collapsed)
        try await Task.sleep(for: .seconds(1))
        XCTAssertEqual(newValue, CardMode.collapsed)
    }
```

**Reply by SebastianBoldt** [reply]
Something that could be interesting is that if I attach a print to the card mode publisher ... 
```swift
static let live: CardModeProvider = {
        return Self(
            cardMode: {
                CardMode(rawValue: UserDefaults.sharedGroupDefaults.cardMode) ?? .expanded
            },
            cardModePublisher: UserDefaults.sharedGroupDefaults
                .publisher(for: \.cardMode,  options: [.initial, .new])
                .print("Cardmode")
                .map { CardMode(rawValue: $0) ?? .expanded }
                .eraseToAnyPublisher(),
            storeCardMode: {
                UserDefaults.sharedGroupDefaults.cardMode = $0.rawValue
            }
        )
    }()
```
I get the following initial output:
```swift
Cardmode: receive subscription: (KVOSubscription)
Cardmode: request max: (1)
Cardmode: receive value: (collapsed)
Cardmode: request max: (1)
```

I am not exactly sure what that max: (1) means but maybe it is a hint.

After I change the value in the UserDefaults again nothing gets printed.

**Reply by tgrapperon** [reply]
That's what I just found too. It seems to work if you interpose `.handleEvents(receiveSubscription: { $0.request(.unlimited) })` between the publisher and `.values`.
```swift
for await value in cardModeProvider
  .cardModePublisher
  .handleEvents(receiveSubscription: { $0.request(.unlimited) })
  .values {
…
}
```
This is likely a behavior that is inherited from the `KVO` publisher, but I'm not knowledgeable enough with Combine to know why it behaves like that in this case. Maybe @iampatbrown has an idea?

**Reply by SebastianBoldt** [reply]
Do you have any Idea why? Is it an Issue on my side?

**Reply by SebastianBoldt** [reply]
There has to be a difference between using eraseToEffect to observe the changes and the for await value in sequence way 🤔

**Reply by tgrapperon** [reply]
I don't think that this is an issue from your side! This is an odd behavior of the KVO Publisher (and maybe other ones) with respect to back pressure I guess. Using `values` on these publishers doesn't seem to produce the same values as sinking on the publisher. This is maybe a bug, or maybe the expected behavior. I'm considering submitting [a PR](https://github.com/pointfreeco/swift-composable-architecture/compare/main...tgrapperon:swift-composable-architecture:publisher-values-doc?expand=1) about this.
A unit test over the `AsyncSequence` would have spotted the issue, but these tests can be a little tricky to set up (as you saw from the combine version). Maybe some test helper for async sequences could help.

**Reply by iampatbrown** [reply]
> Maybe @iampatbrown has an idea?

@SebastianBoldt @tgrapperon I'm pretty sure this is a `KeyValueObservingPublisher` bug similar to [the one](https://forums.swift.org/t/asyncpublisher-causes-crash-in-rather-simple-situation/56574) with `ObservableObjectPublisher`. The workaround is to use `Publishers.Buffer` eg:

```swift
 cardModePublisher: UserDefaults.sharedGroupDefaults
  .publisher(for: \.cardMode,)
  .buffer(size: 1, prefetch: .byRequest, whenFull: .dropOldest)
  .map { CardMode(rawValue: $0) ?? .expanded }
  .eraseToAnyPublisher(),
```

**Reply by tgrapperon** [reply]
Thanks @iampatbrown! This may be a better solution than my `handleEvents` solution because I found that mine is changing the `initial` value behavior and doesn't emit it anymore when we specify `.request(.unlimited)`.
